src/highlight.rs

```diff
@@ -61,17 +61,25 @@ fn prepare_line(s: &str, tab_len: u8, mask: Option<char>) -> Cow<'_, str> {
         None => {
             let tab = spaces(tab_len);
             let mut buf = String::new();
+            let mut col: usize = 0;
             for (i, c) in s.char_indices() {
                 if buf.is_empty() {
                     if c == '\t' {
                         buf.reserve(s.len());
                         buf.push_str(&s[..i]);
-                        buf.push_str(tab);
+                        let len = tab_len as usize - (col % tab_len as usize);
+                        buf.push_str(&tab[..len]);
+                        col += len;
+                    } else {
+                        col += 1;
```

Comment:
```suggestion
                        buf.push_str(tab);
                        col += tab_len;
```

Since the buffer is empty, the inserted tab character is always full-width here.

Reply:
not so. If i type `a tab b` buf is still empty after typing the 'a' character. Buf doesnt get written to until the first tab is encountered, hence the test for `is_empty` at the end. Its a great optimization in the code, 

(I did try it out just to be sure)

Reply:
What about adding some unit tests to show this?

Reply:
there are no tests in tui-textarea that check what the UI looks like. It is doable (I have worked on other projects that use a pseudo backend that renders the screen in memory) but is a huge investment

Reply:
When there's something that seems non-obvious, I generally try to resolve the concern by writing a simple test. It seems reasonably simple to unit test this function. Here's some quick tests generated by copilot and manually tweaked for correctness. (Note to run `cargo test`, you'll need to comment out the line that imports the README into lib.rs as the README contains code that doesn't compile).

These tests pass on main, on the current PR they fail because `tab_len` = 0 causes a panic.

```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_prepare_line() {
        // fn prepare_line(s: &str, tab_len: u8, mask: Option<char>) -> Cow<'_, str> {
        assert_eq!(prepare_line("", 0, None), Cow::Borrowed(""));
        assert_eq!(prepare_line("", 4, None), Cow::Borrowed(""));
        assert_eq!(prepare_line("", 8, None), Cow::Borrowed(""));
        assert_eq!(prepare_line("", 0, Some('x')), Cow::Borrowed(""));
        assert_eq!(prepare_line("", 4, Some('x')), Cow::Borrowed(""));
        assert_eq!(prepare_line("", 8, Some('x')), Cow::Borrowed(""));
        assert_eq!(prepare_line("a", 0, None), Cow::Borrowed("a"));
        assert_eq!(prepare_line("a", 4, None), Cow::Borrowed("a"));
        assert_eq!(prepare_line("a", 8, None), Cow::Borrowed("a"));
        assert_eq!(prepare_line("a", 0, Some('x')), Cow::Borrowed("x"));
        assert_eq!(prepare_line("a", 4, Some('x')), Cow::Borrowed("x"));
        assert_eq!(prepare_line("a", 8, Some('x')), Cow::Borrowed("x"));
        assert_eq!(prepare_line("a\t", 0, None), Cow::Borrowed("a"));
        assert_eq!(prepare_line("a\t", 4, None), Cow::Borrowed("a    "));
        assert_eq!(prepare_line("a\t", 8, None), Cow::Borrowed("a        "));
        assert_eq!(prepare_line("a\t", 0, Some('x')), Cow::Borrowed("xx"));
        assert_eq!(prepare_line("a\t", 4, Some('x')), Cow::Borrowed("xx"));
        assert_eq!(prepare_line("a\t", 8, Some('x')), Cow::Borrowed("xx"));
        assert_eq!(prepare_line("\t", 0, None), Cow::Borrowed("\t"));
        assert_eq!(prepare_line("\t", 4, None), Cow::Borrowed("    "));
        assert_eq!(prepare_line("\t", 8, None), Cow::Borrowed("        "));
        assert_eq!(prepare_line("\t", 0, Some('x')), Cow::Borrowed("x"));
        assert_eq!(prepare_line("\t", 4, Some('x')), Cow::Borrowed("x"));
        assert_eq!(prepare_line("\t", 8, Some('x')), Cow::Borrowed("x"));
        assert_eq!(prepare_line("a\tb", 0, None), Cow::Borrowed("ab"));
        assert_eq!(prepare_line("a\tb", 4, None), Cow::Borrowed("a    b"));
        assert_eq!(prepare_line("a\tb", 8, None), Cow::Borrowed("a        b"));
        assert_eq!(prepare_line("a\tb", 0, Some('x')), Cow::Borrowed("xxx"));
        assert_eq!(prepare_line("a\tb", 4, Some('x')), Cow::Borrowed("xxx"));
        assert_eq!(prepare_line("a\tb", 8, Some('x')), Cow::Borrowed("xxx"));
        assert_eq!(prepare_line("a\t\tb", 0, None), Cow::Borrowed("ab"));
        assert_eq!(prepare_line("a\t\tb", 4, None), Cow::Borrowed("a        b"));
        assert_eq!(
            prepare_line("a\t\tb", 8, None),
            Cow::Borrowed("a                b")
        );
        assert_eq!(prepare_line("a\t\tb", 0, Some('x')), Cow::Borrowed("xxxx"));
        assert_eq!(prepare_line("a\t\tb", 4, Some('x')), Cow::Borrowed("xxxx"));
        assert_eq!(prepare_line("a\t\tb", 8, Some('x')), Cow::Borrowed("xxxx"));
    }
}
```

I'd suggest start with these and then modify them to show what changes in this PR.

Reply:
@pm100 Oh, you're right. The buffer is empty until first tab character so line is not necessarily empty.

@joshka Thank you for the example test case. I'll add some tests after merging this PR.